pallets/file-system/src/lib.rs

```diff
@@ -284,38 +302,60 @@ pub mod pallet {
     where
         u32: TryFrom<BlockNumberFor<T>>,
     {
-        fn on_idle(block: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
+        fn on_idle(current_block: BlockNumberFor<T>, remaining_weight: Weight) -> Weight {
             let db_weight = T::DbWeight::get();
 
-            // Return early if the remaining weight is not enough to perform the operation
-            // TODO: fix this, it's not working as expected (when the remaining weight is 0, it still goes through)
-            if !remaining_weight
-                .all_gte(db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into()))
-            {
-                LastBlockCleaned::<T>::put(block.saturating_add(1u32.into()));
-                return Weight::zero();
+            // Determine the starting block for cleanup, using `NextBlockToCleanup` if available,
+            // or defaulting to the current block
+            let start_block = NextStartingBlockToCleanUp::<T>::get().unwrap_or(current_block);
+            let mut block_to_clean = start_block;
+
+            // Total weight used to avoid exceeding the remaining weight
+            let mut total_used_weight = Weight::zero();
+
+            // Iterate over blocks from the start block to the current block,
+            // cleaning up storage requests until the remaining weight is insufficient
+            while block_to_clean <= current_block && remaining_weight.all_gte(total_used_weight) {
+                // Calculate the weight required to process the cleanup
+                let required_weight =
+                    db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into());
+
+                // Check if there is enough weight left to process the cleanup for one more block
+                if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
+                    break;
+                }
+
+                let mut used_weight = db_weight.reads(1);
+                let expired_requests = StorageRequestExpirations::<T>::take(&block_to_clean);
+
+                // Remove expired storage requests for the block
+                for location in expired_requests {
```

Comment:
For example, in here, it is never checked if there is going to be enough weight left to cleanup all the expired requests (to execute all the `for` loop). Instead, you clear them, register the weight, and only after, in the next `while` loop iteration, you check if you went too far.

Reply:
Before this loop I check if we can clear the potential total maximum of expired storage requests.

```rust
// Calculate the weight required to process the 
let required_weight = db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into());

// Check if there is enough weight left to process the cleanup for one more block
if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
    break;
}
```

Reply:
So there is no chance of exceeding the `remaining_weight` because I check for max.

Reply:
Aaahh good point, yes